### ISSUE TITLE
Fix `no-descending-specificity` false positives for nested rules without declarations

### ISSUE THREAD

**Issue by romainmenke** [stylelint/stylelint#7846]
### What minimal example or steps are needed to reproduce the bug?

```css
.foo {
	& h2 {
		& + p {
			color: blue;
		}
	}
}

h1 {
	& h2 {
		& + #baz {
			color: green;
		}
	}
}
```

### What minimal configuration is needed to reproduce the bug?

```json
{
  "extends": [
    "stylelint-config-standard"
  ]
}
```

### How did you run Stylelint?

[Demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcIB0AzCEAEwA6A7bALgGQYAWATFnoYSQNQYAOV+BNBkANhAE7wYAjTgFcYAbmpsCAX0mzc8vKQCMLWmUo5W6hgGIBAQwBeaqYS69+Acx4wYuCdply80kABoQKAJacYAOQMAWzhEGAAPYMY-JDAAZziPcAhcHysEEC0MDGwQCIJ7KDjc-gBtPGzs3LiCAE8-Tm9cAgBaSFTvKxaag1woAx4oXIqMAF1XJPa0gDFeIIMCDIArOJSk2EZExCyckBr6mEbmkt3OBZga3PcR6rqGptapzu6CXv7Bk9yzgsuQCekgA)

### Which Stylelint-related dependencies are you using?

```json
{
  "stylelint": "latest",
  "stylelint-config-standard": "latest"
}
```

### What did you expect to happen?

No error to be reported

### What actually happened?

> ERROR
Expected selector "& h2" to come before selector ".foo h2" [(no-descending-specificity)](https://stylelint.io/user-guide/rules/no-descending-specificity) [10:2-6]

### Do you have a proposal to fix the bug?

Only declarations are affected by specificity.
Nested rules are not. The combined selector of nested and parent rules do have specificity. But when a parent **rule** only contains other rules it never has any effect by itself.

-----

With nesting it possible to have parent rules with lower specificity and child rules with much higher specificity.

- `.foo h2` has specificity `[0, 1, 1]` -> highest
- `h1 h2` has specificity `[0, 0, 2]`

While the children are the inverse:
- `.foo h2 + p` has specificity `[0, 1, 2]`
- `h1 h2 + #baz` has specificity `[1, 0, 2]` -> highest

We should only warn about specificity when this actually matters, when a rule contains declarations.

-----

[Demo with at rules](https://stylelint.io/demo/#N4Igxg9gJgpiBcIB0AzCEAEwA6A7bALgGQYAWATFnoYQAIC2MUAlgIYYDOYATjDLlXwEaBSABsI3eBgBGYgK4wA3NWEEAvqs25teUgEZBhEhSNqGTNpx58BOISPGTpAc178VDjVrzqQAGhAUZjEYADlWRgQQGAAPSIAHUKQwDg4A8AhcYJdo+wwMbBjYgn4oDiLpAG08AoKijgIAT1CxZlwCAFpIbOYXTsbWXChWbigi2owAXV8MnpyAMUl6VgJogCsOLIzYBPTEfMKQRpaYNo7Ko7FVmEai-0mG5tb2rvm+gYIhkbHLouvSncQLN1EA)

At rules containing declarations should count as declarations.

```css
.foo {
	& h2 {
		@media screen {
			color: blue;
		}
	}
}

h1 {
	& h2 { /* must still warn here */
		@media screen {
			color: green;
		}
	}
}
```

**Comment by romainmenke** [comment]
I also think that this aspect was part of the issue in https://github.com/stylelint/stylelint/issues/5894

I think that if we resolve this issue we can remove the warning not to use this rule with nested CSS:

> We recommend turning this rule off if you use a lot of nesting.

**Comment by ybiquitous** [comment]
@romainmenke Thanks for the report. Sounds great. Is there a blocker to make this issue ready?

**Comment by romainmenke** [comment]
No blockers that I can see :)

**Comment by ybiquitous** [comment]
Okay. I've labeled the issue as **ready to implement**. Please consider [contributing](https://stylelint.io/contributing) if you have time.

There are [steps on how to fix a bug in a rule](https://stylelint.io/developer-guide/rules#fix-a-bug-in-a-rule) in the Developer guide.
